src/main/java/org/openelisglobal/common/rest/provider/NonConformEventsRestController.java

```diff
@@ -0,0 +1,250 @@
+      String ncenumber = String.valueOf(System.currentTimeMillis());
+      NcEvent event = nonConformingEventWorker.create(
+        params.get("labOrderNumber"),
+        Arrays.asList(params.get("specimenId").split(",")),
+        systemUser.getId(),
+        ncenumber
+      );
+
+      eventData.put("nceNumber", event.getNceNumber());
+      eventData.put("id", event.getId());
+
+      Sample sample = getSampleForLabNumber(params.get("labOrderNumber"));
+      if (sample != null) {
+        List<SampleItem> sampleItems = new ArrayList<>();
+        String[] sampleItemIdArray = params.get("specimenId").split(",");
+        for (String s : sampleItemIdArray) {
+          SampleItem si = sampleItemService.getData(s);
+          sampleItems.add(si);
+        }
+        eventData.put("specimens", sampleItems);
+      }
+
+      eventData.put("currentUserId", getSysUserId(request));
+
+      eventData.put(
+        "reportUnits",
+        DisplayListService.getInstance()
+          .getList(DisplayListService.ListType.TEST_SECTION_ACTIVE)
+      );
+      requesterService.setSampleId(sample == null ? null : sample.getId());
+      eventData.put("site", requesterService.getReferringSiteName());
+      eventData.put(
+        "prescriberName",
+        requesterService.getRequesterLastFirstName()
+      );
+      eventData.put("nceCategories", nceCategoryService.getAllNceCategories());
+      eventData.put(
+        "reportDate",
+        DateUtil.formatDateAsText(Calendar.getInstance().getTime())
+      );
+
+      return ResponseEntity.ok(eventData);
+    } catch (Exception e) {
+      return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(
+        "An error occurred while processing the request."
+      );
+    }
+  }
+
+  @PostMapping(
+    value = "/rest/reportnonconformingevent",
+    produces = MediaType.APPLICATION_JSON_VALUE
+  )
+  public ResponseEntity<?> postReportNonConformingEvent(
+    @RequestBody NonConformingEventForm form
+  ) {
+    try {
+      boolean updated = nonConformingEventWorker.update(form);
+      if (updated) {
+        return ResponseEntity.ok().body(Map.of("success", true));
+      } else {
+        return ResponseEntity.ok().body(Map.of("success", false));
+      }
+    } catch (Exception e) {
+      return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(
+        "An error occurred while processing the request."
+      );
+    }
+  }
+
+  private Map<String, Object> addSample(Sample sample) {
+    Map<String, Object> sampleMap = new HashMap<>();
+    sampleMap.put("id", sample.getId());
+    List<Map<String, Object>> sampleItemsList = new ArrayList<>();
+    List<SampleItem> sampleItems = sampleItemService.getSampleItemsBySampleId(
+      sample.getId()
+    );
+    sampleMap.put("labOrderNumber", sample.getAccessionNumber());
+
+    for (SampleItem sampleItem : sampleItems) {
+      Map<String, Object> sampleItemMap = new HashMap<>();
+      sampleItemMap.put("id", sampleItem.getId());
```

Comment:
Same here , Use Stronly typed Java Beans instead of loosely typed Maps for return objects